### PR TITLE
[WIP] FreeBSD support

### PR DESCRIPTION
**Pull request by myfreeweb**
Trying to continue the  `freebsd` branch by @tiehuis… (well, by starting over and cherry-picking from there as needed)

- the [vmovdqa problem](https://lists.llvm.org/pipermail/llvm-dev/2017-September/117860.html) from Windows also appears here
- weird issue with `cache_hash` `cache_hit`: `buf_eql_buf(this_path, chf->path)` fails because the `chf->path` is the same path but 1 character longer (extra zero)
- for now,
  - debug info is not implemented
  - `Dir` `nextFreebsd` is stubbed out
  - only `getrandom` (very recent syscall — FreeBSD 12) is used
  - `path` `realC` uses a Linux-style (`fdescfs`) approach instead of libc
  - `selfExePath` reads `procfs` (should use `sysctl` like the C++ version)
  - libc is used for sysctl and kqueue (copied from darwin)
  - only amd64 is there (I'll do aarch64 once amd64 is done)

Current status (of the self-hosted compiler):

```
Test 1/68 compare-output hello world with libc (Debug)...OK
Test 2/68 compare-output hello world with libc (ReleaseSafe)...OK
Test 3/68 compare-output hello world with libc (ReleaseFast)...OK
Test 4/68 compare-output hello world with libc (ReleaseSmall)...OK
Test 5/68 compare-output multiple files with private function (Debug)...OK
Test 6/68 compare-output multiple files with private function (ReleaseSafe)...Process /usr/home/greg/src/github.com/ziglang/zig/zig-cache/test terminated unexpectedly
error: TestFailed
```

```
* thread #1, name = 'test', stop reason = signal SIGSEGV: invalid address (fault address: 0xfffff80003cd2530)
    frame #0: 0x00000000002014ad test`File_OutStream_writeFn at index.zig:357
   354      const max_bytes_len = 0x7ffff000;
   355
   356      var index: usize = 0;
-> 357      while (index < bytes.len) {
   358          const amt_to_write = math.min(bytes.len - index, usize(max_bytes_len));
   359          const rc = posix.write(fd, bytes.ptr + index, amt_to_write);
   360          const write_err = posix.getErrno(rc);
(lldb) fr v
([]u8) bytes = (ptr = <Bad address>, len = 0)
```

This is a weird bug. That test binary actually completes successfully when run under `truss` (syscall tracing), but does this when run on its own. I have no idea how that is happening…
 
Also, why does the test thing generate test binaries for all the other platforms before the native one?? How do I turn that off?

### CODE REVIEW

**Comment by euantorano** [comment]
Awesome! I started having a look at this but hit some issues on the C++ side of things that I didn't get a hance to look into at all.

**Comment by andrewrk** [comment]
Thanks for doing this work. Is there any way to auto run tests on FreeBSD on master branch pushes?

>     * the `chf->path` is the same path but 1 character longer (extra zero)

This is probably a bug in one of the os.cpp functions, setting an incorrect buffer length.

File: cmake/Findlld.cmake [hunk 1]
```diff
@@ -9,9 +9,13 @@
 find_path(LLD_INCLUDE_DIRS NAMES lld/Common/Driver.h
     PATHS
         /usr/lib/llvm-6.0/include
+        /usr/local/llvm60/include
```
Comment (andrewrk) [inline cmake/Findlld.cmake]: These 6s should be 7s (master branch is incorrect)

File: src/cache_hash.cpp [hunk 2]
```diff
@@ -343,7 +343,8 @@ Error cache_hit(CacheHash *ch, Buf *out_digest) {
             return ErrorInvalidFormat;
         }
         Buf *this_path = buf_create_from_slice(file_path);
-        if (chf->path != nullptr && !buf_eql_buf(this_path, chf->path)) {
+        // XXX: on FreeBSD, chf->path is same text but with extra \0 at the end
```
Comment (andrewrk) [inline src/cache_hash.cpp]: Gotta track this bug down. Figure out where chf->path is getting set, figure out how the buf len gets set incorrectly. Can't just put a bandage on the symptom.

**Comment by euantorano** [comment]
> Is there any way to auto run tests on FreeBSD on master branch pushes?

I had a browse around and there don't seem to be any services like Travis CI that run tests on FreeBSD, so I think the easiest solution might be to just set up Jenkins with a few build agents on various operating systems to cover the BSDs and some other platforms.

**Comment by ararslan** [comment]
A bit tangential but just FYI, the [Julia language repo](https://github.com/JuliaLang/julia) has self-hosted FreeBSD CI set up on a contributor's server, managed via [Buildbot](https://buildbot.net/). If you want to go this route, you can have a look at [our setup](https://github.com/JuliaCI/julia-buildbot/blob/master/master/freebsd_ci.py).

**Comment by myfreeweb** [comment]
Looks like there's already [a buildbot](https://buildbot.git.icu/buildbot/#/) for [linux/aarch64](https://github.com/ziglang/zig/pull/1429)

File: src/cache_hash.cpp (hunk 2 above)
Comment (myfreeweb) [reply src/cache_hash.cpp]: It's the `SplitIterator` leaving a '\0' in the last component. Would it be okay to add `|| byte == 0` to `SplitIterator_isSplitByte`?

**Comment by myfreeweb** [comment]
Okay, testing with `-Dskip-release=true` gets much further, but…

```
Test 402/504 behavior-freebsd-x86_64-Debug-bare packed struct 24bits...Process 1943 stopped
* thread #1, name = 'test', stop reason = signal SIGBUS: hardware error
    frame #0: 0x0000000000217d8e test`behavior-freebsd-x86_64-Debug-bare packed struct 24bits at struct.zig:259
   256          assert(@sizeOf(Foo96Bits) == 12);
   257      }
   258
-> 259      var value = Foo96Bits.{
   260          .a = 0,
   261          .b = 0,
   262          .c = 0,
(lldb) dis
test`behavior-freebsd-x86_64-Debug-bare packed struct 24bits:
    0x217d80 <+0>:   pushq  %rbp
    0x217d81 <+1>:   movq   %rsp, %rbp
    0x217d84 <+4>:   subq   $0x80, %rsp
    0x217d8b <+11>:  xorps  %xmm0, %xmm0
->  0x217d8e <+14>:  movaps %xmm0, -0x10(%rbp)
    0x217d92 <+18>:  leaq   -0x10(%rbp), %rax
    0x217d96 <+22>:  movzbl -0xe(%rbp), %ecx
```

SIMD registers again, just like in the `vmovdqa` issue!

**Comment by myfreeweb** [comment]
So here's a simplified IR of that Zig test case:

```llvm
%Foo96Bits = type <{ i24, i24, i24, i24 }>
@0 = internal unnamed_addr constant %Foo96Bits zeroinitializer, align 1

; @"behavior-freebsd-x86_64-Debug-barepacked struct 24bits"
define i32 @main(i32) unnamed_addr #0 {
Entry:
  %value = alloca %Foo96Bits, align 1
  %1 = bitcast %Foo96Bits* %value to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 1 %1, i8* align 1 bitcast (%Foo96Bits* @0 to i8*), i64 16, i1 false)
  ret i32 0
}

declare void @llvm.memcpy.p0i8.p0i8.i64(i8* nocapture writeonly, i8* nocapture readonly, i64, i1) #4
```

```asm
# %bb.0:                                # %Entry
        xorps   %xmm0, %xmm0
        movaps  %xmm0, -24(%rsp)
        xorl    %eax, %eax
        retq
```

But for a "similar" C program:

```c
#include <string.h>

struct test {
   unsigned int a : 24;
   unsigned int b : 24;
   unsigned int c : 24;
   unsigned int d : 24;
};

static struct test two = { 0, 0, 0, 0 };

int main() {
   struct test one;
   memcpy((void*)&one, (void*)&two, sizeof(struct test));
}
```

Clang emits:

```llvm
%struct.test = type { i24, i24, i24, i24 }

@two = internal global { i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 } { i8 0, i8 0, i8 0, i8 undef, i8 0, i8 0, i8 0, i8 undef, i8 0, i8 0, i8 0, i8 undef, i8 0, i8 0, i8 0, i8 undef }, align 4

define dso_local i32 @main() #0 {
entry:
  %one = alloca %struct.test, align 4
  %0 = bitcast %struct.test* %one to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 4 %0, i8* align 1 getelementptr inbounds ({ i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 }, { i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8, i8 }* @two, i32 0, i32 0), i64 16, i1 false)
  ret i32 0
}

declare void @llvm.memcpy.p0i8.p0i8.i64(i8* nocapture writeonly, i8* nocapture readonly, i64, i1) #1
```

Which compiles fine:

```asm
# %bb.0:                                # %entry
        movq    two+8(%rip), %rax
        movq    %rax, -8(%rsp)
        movq    two(%rip), %rax
        movq    %rax, -16(%rsp)
        xorl    %eax, %eax
        retq
```

**Comment by tiehuis** [comment]
This looks related to #1217. I didnt get any further on this issue.

**Comment by myfreeweb** [comment]
I figured it out. Stack alignment:

```diff
diff --git i/src/codegen.cpp w/src/codegen.cpp
index 5845ff8c..1101e8d9 100644
--- i/src/codegen.cpp
+++ w/src/codegen.cpp
@@ -568,9 +568,9 @@ static LLVMValueRef fn_llvm_value(CodeGen *g, ZigFn *fn_table_entry) {
         maybe_import_dll(g, fn_table_entry->llvm_value, linkage);
     }
 
-    if (fn_table_entry->alignstack_value != 0) {
-        addLLVMFnAttrInt(fn_table_entry->llvm_value, "alignstack", fn_table_entry->alignstack_value);
-    }
+    //if (fn_table_entry->alignstack_value != 0) {
+        addLLVMFnAttrInt(fn_table_entry->llvm_value, "alignstack", 4);
+    //}
 
     addLLVMFnAttr(fn_table_entry->llvm_value, "nounwind");
     add_uwtable_attr(g, fn_table_entry->llvm_value);
```

Now up to test 426…

**Comment by myfreeweb** [comment]
Now behavior tests pass in debug mode…

Anyway, I'd like to discuss the direct syscalls vs libc thing.
I see that Linux support is implemented directly, while Darwin support relies on libc.

I think I'd like to use libc on FreeBSD, because:

- the Linux reasons for avoiding libc (like the diversity of libcs across distros) do not apply here
- implementing direct syscall support for things like threads doesn't sound fun
- implementing syscall wrappers for all CPU architectures doesn't sound fun
- I'd prefer `LD_PRELOAD` hooks that override libc functions to work on all applications, including Zig ones

**Comment by andrewrk** [comment]
> * implementing direct syscall support for things like threads doesn't sound fun
> * implementing syscall wrappers for all CPU architectures doesn't sound fun

These are ok reasons for starting off FreeBSD support with depending on libc (for syscalls only) but are not long term reasons for sticking with libc. I'm ok with adding freebsd support that depends on libc as a first pass and then later going the next step of eliminating the libc dependency.

> * I'd prefer `LD_PRELOAD` hooks that override libc functions to work on all applications, including Zig ones

There is a pattern in the std lib you can see here:

https://github.com/ziglang/zig/blob/b4801186b720ec8e94005a15e13e58741f4548b4/std/os/index.zig#L168-L175

This accomplishes both supporting `LD_PRELOAD` hooks, as well as non-libc static binaries, depending on whether the user links with libc.

There is another strong reason to avoid a libc dependency when possible, and it is that it gives the zig compiler full knowledge of the entire source. This means that Zig can do some program-wide static analysis, such as stack upper bound calculation (see #157).

That said - I am ok with depending on libc as a first pass, and then let's come back to it later and see if we want to take the extra step of eliminating the libc dependency.

**Comment by andrewrk** [comment]
@myfreeweb hey thanks for continuing this effort. I just wanted to let you know that I have installed FreeBSD on a Raspberry Pi and so I will use that to pick up where you left off here and try to get this merged with all tests passing.

**Comment by myfreeweb** [comment]
yeah, sorry, I've been busy with work and stuff. I'm glad that you're interested in FreeBSD!

**Comment by andrewrk** [comment]
I merged master into this branch and resolved the conflicts, and then pushed it to `freebsd2` branch. I'm closing this PR now. Feel free to open additional PRs that are based on the `freebsd2` branch (rather than master). I'll take charge of slowly working on the `freebsd2` branch, getting all the tests passing, getting CI integration going for FreeBSD, and then merging into master.

**Comment by andrewrk** [comment]
See #1759

File: src/util.cpp [hunk 3]
```diff
@@ -47,7 +47,7 @@ bool ptr_eq(const void *a, const void *b) {
 // Ported from std/mem.zig.
 bool SplitIterator_isSplitByte(SplitIterator *self, uint8_t byte) {
     for (size_t i = 0; i < self->split_bytes.len; i += 1) {
-        if (byte == self->split_bytes.ptr[i]) {
+        if (byte == self->split_bytes.ptr[i] || byte == 0) {
```
Comment (andrewrk) [inline src/util.cpp]: What was this for? I'm reverting this change - this is a bandage, but let's fix the actual bug.

Comment (myfreeweb) [reply src/util.cpp]: see beginning of thread

> the `chf->path` is the same path but 1 character longer (extra zero)

Comment (andrewrk) [reply src/util.cpp]: This likely represents a bug in os.cpp having to do with file paths that needs to be fixed. I'll see if I can reproduce it.

Comment (andrewrk) [reply src/util.cpp]: Here was the real fix: https://github.com/ziglang/zig/commit/d770333827549e5dcbb727c7f434eaac9dd433d2#diff-bf5785092d19ff73388acb473305d70aR1459

**Pull request closed without merging**